src/nn/linear.rs

```diff
@@ -46,10 +46,27 @@ pub fn linear<'a, T: Borrow<super::Path<'a>>>(
 
 impl super::module::Module for Linear {
     fn forward(&self, xs: &Tensor) -> Tensor {
-        if let Some(bias) = &self.bs {
-            xs.matmul(&self.ws.tr()) + bias
-        } else {
-            xs.matmul(&self.ws.tr())
-        }
+        xs.linear(&self.ws, self.bs.as_ref())
     }
 }
+
+#[test]
+/// To run this test, generate test tensors:
+///
+/// ```shell
+/// cd tests/linear
+/// python gen-test-data.py
+/// ```
+#[ignore]
```

Comment:
Is this ignore so that the test is not run by default? If that's the case maybe we could check in the repo the npy files so that it can always be run (in which case the size should probably be reduced in order for the resulting files to only be a couple KB).

Reply:
That's correct. I've struggled to find tensors smaller than 512x512 that would consistently reproduce the issue. Perhaps libtorch takes different code paths depending on the size of the tensors? I'm not sure, but that wouldn't surprise me.

With 512x512 tensors, I have this test passing, which shows that the current `matmul()` implementation is close to the PyTorch output, but not equal. Still, this requires about 1MB of test data.
```rust
    assert!(original_output.allclose(&expected_output, 1e-2, 1e-3, false));
    assert!(!original_output.equal(&expected_output));
    assert!(output.equal(&expected_output));
```

Reply:
With some further trial and error / bisecting, I managed to get it down to a 257x4 linear layer and a 2x257 input tensor (which seems to support my guess that libtorch uses different code paths depending on tensor sizes). This adds up to 24k of test data, so I've gone ahead and committed it, and removed the `#[ignore]`.
14c06c2